### ISSUE TITLE
Problems with German keyboard

### ISSUE THREAD

**Issue by masinter** [Interlisp/medley#334]
(reported by guenther.goerz@fau.de)

>  I downloaded the last version and it starts without problems, BUT I have trouble with the keyboard layout.

> Of course, my Mac has a German keyboard, but I can switch to a US keyboard layout which looks rather familiar (see the attached screen shot of the keyboard viewer).  The numbers and letters work, but many of the special character keys don't.  I tried it with the following results:
![image](https://user-images.githubusercontent.com/1116587/118371569-0037a000-b562-11eb-8904-043da335ae12.png)
first row

 §  is  null
 plusminus  is  null

 -  is  null
 _  is  null

 =  is  ' (quote)
 +  is  "

second row

 [  is  null
 {  is  null

]  is  =
 }  is  +

third row

 ;  is  null
 :  is  null

 '  is  null
 "  is  null

 \  is  null
 |  is  null

fourth row

 `  is  null
 ~  is  null

 ,  is  ;
 <  is  :

 .  is  .
 >  is  >

 /  is  -
 ?  is  leftarrow

**Comment by nbriggs** [comment]
I think the problem lies in a combination of what the Mac does, what X11/XQuartz does, what maiko does to try to address the X11 behavior, and what the Medley/Lisp code does.   Since we last looked, XQuartz has had significant version upgrades -- I don't know whether this will have made a difference in the way that the keyboard handling is addressed/addressable at the XQuartz level.

**Comment by nbriggs** [comment]
Given that the German Mac keyboard has different physical keys, it is difficult for anyone here in the US to independently investigate or fix it.  Below are the ABC-QWERTZ and default German keyboards you get on a US style physical keyboard.

![Screen Shot 2021-05-15 at 10 13 19 AM](https://user-images.githubusercontent.com/8138841/118372395-40008680-b566-11eb-9937-06292d73e5ac.png)
![Screen Shot 2021-05-15 at 10 11 09 AM](https://user-images.githubusercontent.com/8138841/118372398-4131b380-b566-11eb-8696-41c486073114.png)

**Comment by masinter** [comment]
@rmkaplan writes in email:
[WHICHKEY.txt](https://github.com/Interlisp/medley/files/6488105/WHICHKEY.txt)
There might be a German keyboard mapping hanging around somewhere.  But you might be able to roll your own.

The attached file WHICHKEY has  function that tells you all the names that are associated with a key that you hold down.  Once you know what Medley receives when you push a given key on the keyboard, you can use KEYACTION to assign whatever meaning you want to the up and down traansitions of that key.

So:
(WHICHKEY)
hold down a key within the next 3 seconds
see the Medley synonyms for that key.

There is another function DOWNP that takes a keyname and tells you whether or not that particular key is depressed (after a 2 second lag).

**Comment by masinter** [comment]
Rainer Joswig added:
```
I use currently another macOS Lispm emulator, which uses the latest Xquartz as the X11 server and which successfully deals with a German keyboard.
It uses a special keyboard mapping and one can use the "dead" keys like ^, ´, `.

Another way one might also want explore is to use a keyboard mapping software on the Mac side. In the past I used 'Keyboard Maestro' for that in some applications.
```

**Comment by masinter** [comment]
i spent some time trying to see if there's anything about XQuartz on OS 10.13 and 
keyboard mappings. No luck so far.

From Guenther:

coming back to Rainer Joswig's remark that Medley runs on Mac OS 11, let me add that although I am still using Mac OS 10.13, I installed the latest update of XQuartz - which shows an incompatible behavior: All the keys I annotated with "null" in my first mail are simply non-existent, i.e. there is no reaction whatsoever in the terminal. Therefore, Ron's suggestion to fix the problem *in* Medley can't work, because those keys don't send any code. As already mentioned, this behavior is different from that of all other applications I tried. So, I think XQuartz ignores the MAc OS keyboard mapping, at least for Umlauts and alt-keys, and that seems to be independent of the Mac OS version.

**Comment by nbriggs** [comment]
XQuartz, Preferences, Input: [*] Follow system keyboard layout.   XQuartz should have seen the German keyboard if that was the system default to start with, but when I switch back and forth between English and German, setting that causes xterm to see the difference.    What it *doesn't* do is change the keycode transmitted, which is what Medley looks at.

**Comment by nbriggs** [comment]
From Prof. Dr. Goerz -- 
<img width="435" alt="mac-de-keyboard" src="https://user-images.githubusercontent.com/8138841/120010976-a3d97500-bf92-11eb-9882-731fe96bc5d1.png">
<img width="436" alt="mac-de-keyboard-shifted" src="https://user-images.githubusercontent.com/8138841/120010166-ac7d7b80-bf91-11eb-8fac-e505098ab7d2.png">
[xev_event_keyboard.txt](https://github.com/Interlisp/medley/files/6561923/xev_event_keyboard.txt)

From which we see that
>keycode 18 is dead circumflex/degree
>keycode 35 is ssharp/question mark
>keycode 58 is less than/greater than

**Comment by masinter** [comment]
I think typing with non-US keyboards can't use the OS Locale-driven keyboard handling.
It's a historical limitation of the architecture of being able to ask KEYDOWNP on a keyboard where there isn't a 1-1 mapping from keypresses and characters generated, not just for shift/control but for combined form / unicode normalization.
We might be able to hook up an alternate Lisp device which accepts a stream of characters rather than a stream of keystrokes, if there isn't one already. It's hard to imagine what to do about TTYIN though.

**Comment by masinter** [comment]
see discussion 8/23 10:55

**Comment by masinter** [comment]
I'm hoping we can make progress on getting Medley Online to work with international keyboards because WE control the configuration of the X-server and VNC and MAIKO and Lisp code.

**Comment by nbriggs** [comment]
We still don't have any control over what the OS sends to the browser and sent on through VNC to the X server and then to us, and I don't think we have any way of setting up the X server based on what the OS settings are.

**Comment by reflektoin** [comment]
Hi, 
I have a keyboard with Finnish layout and experience also similar problems. For example I'm unable to type the backquote "`" which is often used to write Lisp macros.

I've tried using Medley Online with Lubuntu and Firefox. And running it in a docker container on a server and connecting to it using ssvnc and xtightvncviewer.

**Comment by nbriggs** [comment]
I've got an issue open for XQuartz [246](https://github.com/XQuartz/XQuartz/issues/246) for the fact that XGetKeyboardMapping does NOT return the correct keyboard mapping when the XQuartz option "follow system keyboard layout" is set and the keyboard layout is not the physical one of the mac it's running on (e.g., German system keyboard layout on a US MacBook Pro).

**Comment by masinter** [comment]
This was discussed on LispCore mailing list https://groups.google.com/g/lispcore/c/BTaSEkzaqN4/m/HwHlSdr-AQAJ
https://groups.google.com/g/lispcore/c/BTaSEkzaqN4/m/PQ5tx7THAQAJ

**Comment by farvardin** [comment]
same here with the French layout ("azerty").

I don't know much about Lisp or interlisp. I've tried to launch medley and discovered while most of the keys are correct (a=a, q=q), the upper keys are like on the qwerty keyboard: if I type &é"'( i'll get 12345. To get the () (lowkey 5 and lowkey _ (the key on the right to 9), on azerty, I have to type like on qwerty: shift+9 and shift+0.

**Comment by masinter** [comment]
@farvardin is your problem on a Mac? This isn't a lisp or interlisp problem at its heart -- the problem is with X key maps sand (on the Mac) XQuartz.

**Comment by farvardin** [comment]
@masinter oups sorry, forgot to describe the OS, it's linux mint 21 and i'm using the official deb package

**Comment by nbriggs** [comment]
@farvardin - this all comes about because (for historical reasons with the D-machines and raw access to the Sun keyboards) the Maiko X11 keyboard handling uses the X key code (rather than the key symbol), so it does not get the national keyboard layout applied -- that is left up to LISP code mapping from the physical key to the key symbol.

We don't have good international keyboard support in the LISP code, and we don't have a principled way of passing what X and any configured OS information may have about which national keyboard layout is in play up to the Maiko C code and then to the LISP code.

**Comment by masinter** [comment]
couldn't we provide a table that users could modify and share? consult it at startup time (people rarely change keyboards mid-session)...

**Comment by nbriggs** [comment]
@masinter - one would have *thought* that the X11 keyboard mapping table would be such a table -- but that is not working with XQuartz.  But, we need not only the indication of what national keyboard is in use, but LISP code that sets up the appropriate mappings from the key (bit) number to the Lisp key names.
The way to get the desired outcome is NOT by making random incremental changes to the code to try to bypass the existing mechanisms

**Comment by nbriggs** [comment]
@masinter - if someone wanted to write a piece of C code that can reliably interrogate X11 on macOS/FreeBSD/Linux and come up with the keyboard layout information that we need, that would be very helpful (and can be done independently of the existing C code)

**Comment by rmkaplan** [comment]
On the Lisp side, Virtualkeyboards and other *keyboard* files in /library provide some interfaces for setting up, visualizing, and installing different keyboard mappings into Lisp’s keyaction tables.

So maybe creating an X keyboard configuration would move part way.  I thought the problem, though, was that there are some keys on the German etc. keyboards that are dead with respect to what Lisp sees, don’t result in any bits changing in the keyboard bit table.

**Comment by nbriggs** [comment]
@rmkaplan yes, there are some keys that are dead, though I suspect it's mostly keys that aren't used on the US keyboard (like cursor keys) but we can adjust that if the C code knows the peculiarities of the key codes for the keyboard that is present, and can pass the right information to Lisp such that it can install the correct keyboard file into the keyaction table.

**Comment by simon-brooke** [comment]
UK English keyboard (under Ubuntu, but I'm guessing all Linux distros) also has minor issues – not ones I can't work around, but it would be nice to have them fixed. And I agree, a properly engineered fix is to be preferred. So the keyboard in use needs to be fetched from the host operating system at Lisp startup time, I think, and used to select the appropriate virtual keyboard.

This is an issue I'd be interested in looking at as a sort of 'get my feet wet' issue, if someone could give me a few pointers.

**Comment by nbriggs** [comment]
@simon-brooke - in the Maiko VM code it's pretty messy.   But, if you want to figure out how, with common code between Linux, macOS, FreeBSD, and Solaris, with local X11 display and also with running it via SSH with remote display, you can determine what keyboard is in use... I'd be happy to talk about it -- but in detail only once I'm back from a family obligation, after April 15.  I have limited time until then.

**Comment by simon-brooke** [comment]
@nbriggs - OK, it seems to me that what we need to do is identify a token representing the keyboard layout in the host environment, perhaps normalise that, set a global variable in Medley to that token; then have a hash-map which relates that token to an appropriate Medley keyboard layout.

But I note @masinter's comment (above) that the keyboard handling code isn't clean, so I agree this may be a rabbit hole of uncertain depth. Still, if we want a portable Medley, at some stage it does have to be tackled.

**Comment by simon-brooke** [comment]
I had assumed that there would be a well known open source cross platform library to detect keyboard layouts, but after careful Google search I cannot find one. There doesn't even seem to me to be a standard technique for Linux; although 

> setxkbmap -print -verbose 10

will tell me I have both British and Icelandic keyboard layouts registered I cannot see that it tells me which one is currently selected. Neither GTK nor KDE appear to have a function to get the current keyboard layout.

Windows has a [GetKeyboardLayout](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getkeyboardlayout) library call, which returns an [HKL value](https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types) but there appears to be no documentation on what the values are; however, I suspect they may be the same as the 'keyboard identifier' values [given here](https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/windows-language-pack-default-values?view=windows-11).

For Mac OS, there's a method for [getting the current keyboard layout name](https://gist.github.com/pkamb/6128775) but needless to say these names are not the same as those used by either MS Windows or X.org.

So it seems to me that what's needed is a C library with conditional compilation for each supported environment and an enormous lookup table to standardise the values returned to system-independent values. It would be useful not merely to Medley but to many other projects; but it would be a lot of work and an absolute pig to maintain.

**Comment by masinter** [comment]
Interlisp-10 took input directly from the 'terminal'. Interlisp-D has a 'chatserver' lispusers package or two which take input from a network stream, without any X or keymaps or bitvectors of keydownp.

Some people will want to use a tablet without a keyboard. Rather than maintaining the current system of communicating via key number-indexed keydownp's, couldn't we switch around the translation so that keydownp(keyname) is true if the input stream's last value was keyname's code within 250 milliseconds?

Another problem is the mismatch between Lisp's view of caps-lock and the host oS's view, which can easily get out of sync.

**Comment by nbriggs** [comment]
@simon-brooke and so you see why we don’t have this… and setxkbmap does not work on macOS even for a query.  I think one has to look at what the code and symbol mappings are in effect to deduce which keyboard layout is in use , but again the mapping query does work on macOS

**Comment by fghalasz** [comment]
For the x-windows implementation(s) of maiko, why don't we switch from key-codes to key-symbols?  I think this would radically simplify the problem for us and would actually solve many of the nationality problems we are having since they would be handled by the OS. (I am assuming XQuartz on MacOS handles national keyboards correctly - it just doesn't give you any information about how it does the mapping.)

The maiko keyboard code would need to select/map key-symbols to a 112-bit index (as it does currently for key-codes).  The Lisp code would then map the 112-bit index to Lisp "actions" as it does now.   This would require a pair of matched select/map tables - the key_symbol->112-index and the 112-index->action.   The second table is just a modified version our current Key->Action table in Lisp.  The first would be created in Lisp to match the modified Key->Action table but it would be handed down to maiko.

This way we could create different nationality keyboard setups (i.e., the two aforementioned table) in Lisp that would select which key-symbols we are interested (e.g., umlaut in German but not in US English) without worrying about where the various keys are on the keyboard.

Special handling would have to made on both sides for the various shift/ctrl/meta keys (as happens currently) without having to know the keycodes of e.g., the ctrl and meta keys.

One wrinkle is that the naïve symbol->112-index mapping is inefficient relative to the keycode->112-index mapping since almost every keycode represents two symbols (e.g., "s" and "S") - so would probably need to reserve one-bit of the 112-index to be a pseudo-shift key, that allows each of the other 111 bits to represent two different key symbols. 

This scheme would also solve some of the shift issues, since shift interpretation relative to key symbols would be handled by X Windows.   It would not handle the issues with things like shift-selection since the asynchrony problems in this case are caused by running in a window a system that thinks it owns the whole screen.

Also I am not sure how much of the online keyboard issues are handled by this scheme due the browser->noVnc->Xvnc->maiko complexity.  That'll take a bit more thought.

For non-X11 systems, we would just leave the key mapping system as it is.  Lisp would have to recognize whether it was on X11 or not to adjust the Lisp Key->Action table appropriately.

Thoughts?

**Comment by LdBeth** [comment]
@fghalasz isn't the current key mapping approach already symbol based?

https://github.com/Interlisp/maiko/blob/900c9557ab3a50d7999907ca9fc0827086a4348b/inc/XKeymap.h#L62

I can confirm that running something like `$ xmodmap -e "keycode 9 = l L l L lstroke Lstroke lstroke"` before
launch medley which indeed change the key that sends keycode 9 (in my case it is `s` key) to `l`.

**Comment by nbriggs** [comment]
@LdBeth - Yeah, see `make_X_keymap()` in initkbd.c -- where it uses the XKeymap.h data to attempt to construct keycode to Lisp bit-number mappings based on what symbols it finds.  I'm surprised that an xmodmap hack works in XQuartz, though.

**Comment by fghalasz** [comment]
In the end maiko uses keycodes and not key symbols to process incoming keystrokes.  True during initialization it uses XGetKeyboardMapping and the key symbols therein to set up a mapping table between keycodes and lisp key bit numbers.  But thereafter, key symbols are ignored.  If XGetKeyboardMapping returns accurate information (which I gather under XQuartz it does not), this scheme can handle a portion of the foreign keyboard problems we are having - like the location of the "Z" and "Y" keys on the keyboard.  

But the current scheme cannot handle other problems - e.g., the interpretation of SHIFT-2.  On the US Keyboard, SHIFT-2 results in the @ key symbol (e.g., XK_at) but on the German keyboard SHIFT-2 results in the " key symbol (XK_quotedbl).  But in keycode terms 2 and SHIFT-2 are the same keycode - hence maiko just reports that HK_2 (and HK_Shift_L) are pressed and leaves it up to Lisp to determine what the SHIFT-2 really means - which its in a really poor position to do since it doesn't really have any access to what keyboard is being used and doesn't have a large database of keyboard layouts to rely on..  But the OS and X11 do have this info and layout database and have already figured out what SHIFT-2 is - its in the key symbol.

So my point is that maiko should process key symbols and not key codes- on every keystroke.  The mapping should be from keysym to lisp key  bit numbers.

**Comment by fghalasz** [comment]
As a side note: maiko's XKeyboard.h is the kind of convoluted, non-transparent, "it kind of works but I have no idea how" code that would never be allowed in any decent production system.  Its pure crap coding - no ifs about it.

**Comment by nbriggs** [comment]
@fghalasz - I agree that using key symbols would fix issues with the coding of the keys that are common between the layouts (Z, Y).  I'm not clear on what the Lisp code should expect for the symbol-to-keybit mapping for everything else.

The SDL version of the input is currently set up to use key symbols instead of key codes, check out and compile the "sdl" branch of maiko to get a sense of how that feels.  It has its own key map in sdl.c, which suggests a mapping for 109 of the 112 bits, but Peter didn't say what the criteria he used for the mapping was.

**Comment by masinter** [comment]
I think that's a problem with lots of old code bases which go into "maintenance". Some parts of the system are not well understood, and you get waves of patching, advising, and modifying things at the wrong layer. The 1992 AAAI show [photo](https://drive.google.com/file/d/1SzLbnS4zVM4GroY48IPy_3IlANh3d0au/view?usp=sharing) "Tools to move you .... From a quick prototype thru a finished product".
Error 33: build your research prototype on top of someone else's research prototype.

As for the keyboard, I'd look for a method of receiving keys from alternate input devices as a stream of characters, (not key syms or key numbers), and then try to fake out those who currently look at the bit-vector.   Or keep the old code around for bit-vector code we somehow can't modify, but just note it doesn't necessarily work unless you have a US standard keyboard.

speech to text, hand writing, touch screen, mobile devices. I want a headless medley server that takes type in from STDIN

**Comment by nbriggs** [comment]
@masinter
Using, say, CL:OPEN -- 
```
(setq stdin (open "{unix}/dev/stdin" :direction :input))
(read stdin)
```
This reads ASCII (or UTF-8) input independent of the keyboard setup so you can write a headless medley server trivially.

**Comment by masinter** [comment]
> The SDL version of the input is currently set up to use key symbols instead of key codes, check out and compile the "sdl" branch of maiko to get a sense of how that feels. It has its own key map in sdl.c, which suggests a mapping for 109 of the 112 bits, but Peter didn't say what the criteria he used for the mapping was.

I've been running the SDL branch for ~5 days and have had no problems. It sounds like this might be a big step in solving a number of keyboard issues.

**Comment by lassik** [comment]
The SDL version (`ldesdl`) now works wonderfully. But I also have a Finnish keyboard and have problems with shifted keys on the number row to type punctuation.

As a native user (or sufferer :-) of "internationalization", I strongly recommend feeding characters/keysyms directly into Maiko instead of raw keycodes. The translation will always be complex, buggy, and incomplete unless each backend (X, SDL, web browser, text terminal, ...) is allowed to fully utilize its native knowledge of the keyboard state and layout.

@masinter wrote:

> As for the keyboard, I'd look for a method of receiving keys from alternate input devices as a stream of characters, (not key syms or key numbers), and then try to fake out those who currently look at the bit-vector.

That sounds like the right call.

**Comment by lassik** [comment]
It would be best for maintainability and compatibility with multiple backends if modifier keys (shift, control, etc.) are tracked separately from other keys.

Concretely, Maiko would send Lisp an event stream like this:

```
(left)
(left shift)
(left control shift)
(up)
("a" control)
("ä")  ; international
```

The first element of an event is either:

* a string (denoting a character typed according to the backend's conventions)
* a symbol (denoting a special key such as an arrow key or tab)

Any remaining elements are the set of modifier keys pressed while the event occurred. In C these could be a bitmask.

**Comment by lassik** [comment]
BTW how many bits wide is the I/O region that is used to send the keycode? Unicode is a 21-bit character set. I noticed that the Medley character set designed by @jcowan is 16-bit. Extra I/O bits might be used to encode the state of modifier keys?

**Comment by nbriggs** [comment]
Back up to my previous comment:  There are 112 bits (we can get a few more), but each one represents a key, and Lisp code *does not* work solely at the character level - *every* key can be examined in combination to see if it is in an up or down state.  IF you let the Lisp transition handling code get as far as generating characters then it is generating them into a 16-bit space, as Xerox Character Code standard internal representations.

**Comment by lassik** [comment]
If Maiko itself stores the entire keyboard state as a bitmap in the I/O region, then a new mechanism is probably needed to send international characters from Maiko to Lisp.

Beyond "international characters", we only have English characters and a few special keys. Could the Lisp side assume that those always have their US English keycodes, and Maiko would translate from backend keysyms to those?

**Comment by nbriggs** [comment]
**Lisp** stores the entire keyboard state in the I/O region, Maiko undoes the OS encoding of keys to characters in as much as it knows how, and then sets the bits in Lisp's keyboard state, which Lisp then interprets.  See the code in medley/sources/LLKEY.

In the internal string encoding (which is Xerox Character Code Standard, 16-bit, predates Unicode) Lisp can and does encode international characters.  There's Lisp code for virtual keyboards that lays an international keyboard mapping over the physical key bits - you "just" need to set the mapping correctly.   The problems mostly happen when Maiko doesn't correctly undo the OS's encoding and/or the user is using a US keyboard layout in Medley with a non-US keyboard.

**Comment by lassik** [comment]
I understand the appeal of handling all input duties Lisp-side, but I believe that approach cannot be made to work very reliably. The point of using physical keycodes instead of keysyms is precision, but with so many layers of abstraction underneath Maiko it can only be precise under ideal conditions.

If it's important to retain the physical keycodes (at least under some conditions, e.g. running Maiko without an OS), I will have to rethink my proposal.

**Comment by masinter** [comment]
SDL sends key names which are translated to Lisp keycodes in maiko/src/sdl.c
Doesn't that mean that SDL-based backends will work with international  keyboards?

**Comment by pamoroso** [comment]
The SDL backend does indeed seem to recognize international symbols. Here is the console output of Medley SDL, under Crostini Linux on my Chromebox with Italian keyboard, when I press the corresponding keys at the Exec:

```
paoloamoroso@penguin:~$ medley -a -e -n &
[1] 1397
paoloamoroso@penguin:~$ running: lde  -noscroll -g 1440x900 -sc 1440x900 -title "Medley Interlisp" -m 256  -id default /home/paoloamoroso/medley/medley/loadups/apps.sysout
greet: /home/paoloamoroso/medley/medley/greetfiles/APPS-INIT.LCOM
requested width: 1440, height: 900
initialised
Window created
Creating renderer...
Creating texture...
SDL initialised
Unsupported keyboard type: 14
Configuring keyboard for type-3
No mapping for key Up
No mapping for key Up
No mapping for key Down
No mapping for key Down
No mapping for key Left
No mapping for key Left
No mapping for key Right
No mapping for key Right
No mapping for key à
No mapping for key à
No mapping for key è
No mapping for key è
No mapping for key è
No mapping for key è
No mapping for key ì
No mapping for key ì
No mapping for key ò
No mapping for key ò
No mapping for key ù
No mapping for key ù
```

However, none of the symbols appear in the Exec or carry out the corresponding action. In addition, each symbol is printed twice on the console even if I press the key once.

**Comment by nbriggs** [comment]
@masinter - it works differently from the X11 version and the mapping is incomplete.  It will be more likely to have the character printed on the key show up in Lisp, at least when it's characters that are on the English keyboard and are known by Lisp -- e.g., with a German QWERTZ keyboard you can probably type the Z (and the parens which are 8/9 instead of 9/0) -- but if you type o-umlaut or a-umlaut from that keyboard (which are where semicolon and quote are on a US keyboard), for one thing you'll get `No mapping for key ä` (because the table doesn't include one) and since as far as Lisp is concerned there's no a-umlaut physical key it's unclear what the mapping to keyboard bit should be anyway.

**Comment by nbriggs** [comment]
@pamoroso - there's one line printed for the key-down action, another for the key-up action.  Nothing gets to the exec because the maiko code receiving the key actions from SDL prints the "No mapping" message and ignores the key... if there's no entry in the mapping table.

**Comment by masinter** [comment]
I vaguely remember a post from John Cowan about a design that would leave the current bit table except one bit which was "something else" which would, mean that the key code was in another field.
We could use the Unicode <-> XCCS conversion routine. ...

**Comment by nbriggs** [comment]
Yes, we have had a number of partial designs, however I haven't yet seen one that lays out exactly how it would work from the maiko key event reading (from X11 or SDL or ...) through to, and including, the keyboard event handling in Lisp in a way that can get a decoded character into a buffer without confusing the keyboard event handling code.

**Comment by lassik** [comment]
Would it be acceptable to send international characters using some kind of queue? The character code could be encoded directly into the queue by Maiko. Lisp would not assume that it corresponds to one physical key.

**Comment by lassik** [comment]
Perhaps simplest would be to keep the input character queue on the C side?

In the I/O region, there would be one word which is zero initially and into which one international character at a time is removed from the queue.

Lisp checks the I/O word. If nonzero, it handles the character. It writes back zero to the same address to indicate that it has consumed the character.

Upon noticing that Lisp has zeroed the I/O word, Maiko would remove the next character from the queue and place it in the I/O word. If the queue is empty, the word remains zero until there is new input again.

The C-side queue should have a limit as to how many characters it stores. It could also have a timeout so that stale input is dropped.

This queue would only concern non-ASCII characters.

**Comment by nbriggs** [comment]
`INTCHARCODE` is for when the character is defined as an interrupt generating character.  The macros are related to the entries in a keyaction table.

**Comment by masinter** [comment]
merge with issue #1780 (closing this one; #1780 has more recent info)
